### PR TITLE
Optimize batch insertion of edges

### PR DESCRIPTION
**Pull request by ehein6**
## Summary
Provides a new family of functions for inserting an entire batch of edge updates into stinger at once. Because they perform multiple updates for each edge list traversal, these functions can be much faster (~10x) than the current implementation, which traverses the edge list at least once for each update. 

## Algorithm Overview
1. Sort the list of edge updates by source vertex ID
2. Each thread grabs a sub-list of updates for a single source vertex ID and walks that vertex's edge list
3. For each edge encountered, it performs a binary search on its sublist for any matching updates. Empty slots are filled with the next available edge in the list.

## API
The new functions are prefixed with `stinger_batch_`, and their functionality corresponds to the existing functions for inserting edges:
- `stinger_insert_edge` -> `stinger_batch_insert_edges`
- `stinger_incr_edge` -> `stinger_batch_incr_edges`
- `stinger_insert_edge_pair` -> `stinger_batch_insert_edge_pairs`
- `stinger_incr_edge` -> `stinger_batch_incr_edge_pairs`

The functions accept a pair of [iterators](http://en.cppreference.com/w/cpp/concept/RandomAccessIterator) to the range of updates to perform, which will be sorted in-place. The caller must also supply a set of adapter functions for getting/setting the source, destination, weight, type, and result code for an edge update as a template parameter. This design allows the functions to be independent of the underlying edge update type (e.g. `EdgeInsertion` or `stinger_edge_update`)

## Integration
The stinger batch server has been upgraded to use the new functions.

---
This change is [Reviewable](https://reviewable.io/reviews/stingergraph/stinger/208)

### CODE REVIEW

**Comment by ehein6** [comment]
This is NOT ready to merge yet. Pending issues:

- [ ] Rare deadlock bug
- [ ] Add unit tests for undirected edges
- [ ] Add unit tests for multiple edge types
- [ ] Performance tuning (pick threshold for reverting back to old algorithm)

**Comment by ehein6** [comment]
This should help a lot with issue #170.

**Comment by ehein6** [comment]
Depends on #209

File: lib/stinger_core/inc/stinger_batch_insert.h [hunk 1]
```diff
@@ -0,0 +1,571 @@
...
+            int64_t neighbor = use_dest::get(*pos);
+            int64_t weight = adapter::get_weight(*u);
+            int64_t time = adapter::get_time(*u);
+            if (u == pos) {
+                adapter::set_result(*u, result);
+                update_edge_data_and_direction (G, eb, e, neighbor, weight, time, direction, create ? EDGE_WEIGHT_SET : operation);
+            } else {
+                adapter::set_result(*u, EDGE_UPDATED);
+                update_edge_data_and_direction (G, eb, e, neighbor, weight, time, direction, operation);
+            }
```
Comment (tgoodyear) [inline lib/stinger_core/inc/stinger_batch_insert.h]: Would it make sense to convert this to a canonical for loop and parallelize it? What is an edge slot?

File: lib/stinger_core/inc/stinger_batch_insert.h [hunk 2]
```diff
@@ -0,0 +1,571 @@
...
+            switch (result){
+                case EDGE_ADDED:
+                case EDGE_UPDATED:
+                case EDGE_NOT_ADDED:
+                    // Subtract 10 to get back to return code caller expects
+                    adapter::set_result(*u, result - 10);
+                    break;
+                case -1:
+                case 1:
+                    // Caller must have set the return code, leave it be
```
Comment (tgoodyear) [inline lib/stinger_core/inc/stinger_batch_insert.h]: Can we define these magic ints alongside `EDGE_ADDED`, etc?

File: lib/stinger_core/inc/stinger_batch_insert.h [hunk 3]
```diff
@@ -0,0 +1,571 @@
...
+        clear_results(pos, updates_end);
+        LOG_V("Beginning OUT updates for second half of batch...");
+        do_batch_update<OUT, source_funcs, dest_funcs>(G, pos, updates_end, operation);
+
+        // For undirected, do the same updates again in the opposite direction
+        if (!directed)
+        {
+            // TODO does anyone really care what happens when half of an undirected insert has a different return code?
+            clear_results(updates_begin, updates_end);
+
```
Comment (tgoodyear) [inline lib/stinger_core/inc/stinger_batch_insert.h]: How does this "half of an undirected insert has a different return code" situation happen?

File: lib/stinger_core/inc/stinger_batch_insert.h [hunk 4]
```diff
@@ -0,0 +1,571 @@
...
+     * pos - pointer to first update
+     * updates_end - pointer to end of update list
+     * G - pointer to STINGER
+     * eb - pointer to edge block
+     * e - edge index within block
+     * operation - EDGE_WEIGHT_SET or EDGE_WEIGHT_INCR
+     */
+    template<int64_t direction, class use_dest>
+    static void
+    do_edge_updates(int64_t result, bool create, iterator pos, iterator updates_end, stinger_t * G, stinger_eb *eb, size_t e, int64_t operation)
```
Comment (tgoodyear) [inline lib/stinger_core/inc/stinger_batch_insert.h]: Long line

File: lib/stinger_core/inc/stinger_batch_insert.h [hunk 5]
```diff
@@ -0,0 +1,571 @@
...
+                    // Ran out of edge blocks!
+                    writeef (block_ptr, (uint64_t)old_eb);
+                    while(next_update() != updates_end)
+                    {
+                        adapter::set_result(*next_update(), EDGE_NOT_ADDED);
+                    }
+                    return;
+                } else {
+                    // Add our edge to the first slot in the new edge block
+                    do_edge_updates<direction, use_dest>(EDGE_ADDED, true, next_update(), updates_end, G, ebpool_priv + newBlock, 0, operation);
```
Comment (tgoodyear) [inline lib/stinger_core/inc/stinger_batch_insert.h]: Long line

File: src/server/src/batch.cpp [hunk 6]
```diff
@@ -109,6 +110,202 @@ handle_edge_names_types(EdgeInsertion & in, stinger_t * S, std::string & src, st
...
+    if (batch.make_undirected())
+    {
+        stinger_batch_incr_edge_pairs<EdgeInsertionAdapter>(
+                S, batch.mutable_insertions()->begin(), batch.mutable_insertions()->end());
+    } else {
+        stinger_batch_incr_edges<EdgeInsertionAdapter>(
+                S, batch.mutable_insertions()->begin(), batch.mutable_insertions()->end());
+    }
+
+    OMP("omp parallel for")
```
Comment (tgoodyear) [inline src/server/src/batch.cpp]: Any reason to use static over dynamic OMP scheduling here?

File: src/tests/stinger_batch_test/stinger_batch_test.cpp [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+    // Weight should be sum of all updates
+    STINGER_FORALL_EDGES_OF_ALL_TYPES_BEGIN(S) {
+        EXPECT_EQ(STINGER_EDGE_WEIGHT, total_weight_per_edge);
+    }STINGER_FORALL_EDGES_OF_ALL_TYPES_END();
+
+    for (int i=0; i < 100; i++) {
+        EXPECT_EQ(stinger_outdegree_get(S, i), 99 - i);
+    }
+}
+
```
Comment (tgoodyear) [inline src/tests/stinger_batch_test/stinger_batch_test.cpp]: Should we add tests that also include deletions?

File: lib/stinger_core/inc/stinger_batch_insert.h [hunk 8]
```diff
@@ -0,0 +1,571 @@
...
+        static bool
+        compare(const update &a, const update &b){
+            return adapter::get_dest(a) < adapter::get_dest(b);
+        }
+        static bool
+        sort(const update &a, const update &b){
+            return (adapter::get_type(a) != adapter::get_type(b)) ? adapter::get_type(a) < adapter::get_type(b) :
+                   (adapter::get_dest(a) != adapter::get_dest(b)) ? adapter::get_dest(a) < adapter::get_dest(b) :
+                   (adapter::get_source(a) != adapter::get_source(b)) ? adapter::get_source(a) < adapter::get_source(b) :
+                   (adapter::get_time(a) != adapter::get_time(b)) ? adapter::get_time(a) < adapter::get_time(b) :
```
Comment (ehein6) [inline lib/stinger_core/inc/stinger_batch_insert.h]: No, this one orders by dest then source, the other one orders by source then dest.

File: lib/stinger_core/inc/stinger_batch_insert.h [hunk 9]
```diff
@@ -0,0 +1,571 @@
...
+        static bool
+        compare(const update &a, const update &b){
+            return adapter::get_source(a) < adapter::get_source(b);
+        }
+        static bool
+        sort(const update &a, const update &b){
+            return (adapter::get_type(a) != adapter::get_type(b)) ? adapter::get_type(a) < adapter::get_type(b) :
+                   (adapter::get_source(a) != adapter::get_source(b)) ? adapter::get_source(a) < adapter::get_source(b) :
+                   (adapter::get_dest(a) != adapter::get_dest(b)) ? adapter::get_dest(a) < adapter::get_dest(b) :
+                   (adapter::get_time(a) != adapter::get_time(b)) ? adapter::get_time(a) < adapter::get_time(b) :
```
Comment (ehein6) [inline lib/stinger_core/inc/stinger_batch_insert.h]: How would you make it prettier? I think this is still better than a quadruply-nested if statement.

File: lib/stinger_core/inc/stinger_batch_insert.h (hunk 4 above)
Comment (ehein6) [inline lib/stinger_core/inc/stinger_batch_insert.h]: Agree, will fix.

File: lib/stinger_core/inc/stinger_batch_insert.h (hunk 1 above)
Comment (ehein6) [inline lib/stinger_core/inc/stinger_batch_insert.h]: This is already inside a larger parallel for, so it doesn't make sense to parallelize again. 

An edge block consists of a header followed by a fixed-size array (default 14) of edges. I called it an edge slot because its really just a place to store an edge; if we delete the edge or haven't allocated one yet, it's just an empty space.

File: lib/stinger_core/inc/stinger_batch_insert.h (hunk 5 above)
Comment (ehein6) [inline lib/stinger_core/inc/stinger_batch_insert.h]: Will wrap

File: lib/stinger_core/inc/stinger_batch_insert.h (hunk 2 above)
Comment (ehein6) [inline lib/stinger_core/inc/stinger_batch_insert.h]: Unfortunately the calling code uses magic numbers too, but I suppose I could add these to the enum.

File: lib/stinger_core/inc/stinger_batch_insert.h (hunk 9 above)
Comment (tgoodyear) [inline lib/stinger_core/inc/stinger_batch_insert.h]: My preference would be to define a function that looks like the following. No need to nest. I could be alone in this opinion.

```
int sort_source_dest(const update &a, const update &b){
	if (adapter::get_type(a) != adapter::get_type(b))
		return adapter::get_type(a) < adapter::get_type(b);
	
	if (adapter::get_source(a) != adapter::get_source(b)) 
		return adapter::get_source(a) < adapter::get_source(b);

	if (adapter::get_dest(a) != adapter::get_dest(b))
		return adapter::get_dest(a) < adapter::get_dest(b);

	if (adapter::get_time(a) != adapter::get_time(b))
		return adapter::get_time(a) < adapter::get_time(b);
	return false;
}
```

File: src/tests/stinger_core_test/stinger_core_test.cpp [hunk 10]
```diff
@@ -365,6 +365,8 @@ TEST_F(StingerCoreTest, stinger_edge_insertion) {
   EXPECT_EQ(consistency,0);
 }
 
+
+
```
Comment (tgoodyear) [inline src/tests/stinger_core_test/stinger_core_test.cpp]: Unnecessary whitespace addition

File: lib/stinger_core/inc/stinger_batch_insert.h (hunk 3 above)
Comment (ehein6) [inline lib/stinger_core/inc/stinger_batch_insert.h]: An undirected edge is implemented as one directed edge in each direction. The existing implementation in `stinger_insert_edge_pair` concatenates the return codes from both edge inserts in a single int. For example, if the graph already has a directed edge from 4->5, and we try to insert an undirected edge 4<->5. Edge 4->5 returns 0 (edge exists), 5->4 returns 1 (edge added), this is concatenated into 0x2. 

My code doesn't replicate this behavior. It would be tricky to implement and doesn't seem useful to me. Typically graphs will be all directed or undirected edges, so we expect both inserts to return the same value. But I could fix this if necessary.

File: src/server/src/batch.cpp (hunk 6 above)
Comment (ehein6) [inline src/server/src/batch.cpp]: This loop is just checking return codes and optionally doing vertexID -> name lookups. Each iteration ought to do about the same amount of work, so a static schedule will be balanced with low synchronization overhead.

File: src/tests/CMakeLists.txt [hunk 11]
```diff
@@ -10,16 +10,27 @@ target_link_libraries(stinger_names_test stinger_core gtest)
 #================================
 
 set(_stinger_core_test_sources
-  stinger_core_test/stinger_core_test.cpp
-  stinger_core_test/stinger_core_test.h
-)
+        stinger_core_test/stinger_core_test.cpp
+        stinger_core_test/stinger_core_test.h
+        )
```
Comment (ehein6) [inline src/tests/CMakeLists.txt]: Oops, need to fix whitespace here.

File: src/tests/CMakeLists.txt [hunk 12]
```diff
@@ -10,16 +10,27 @@ target_link_libraries(stinger_names_test stinger_core gtest)
...
 add_executable(stinger_core_test ${_stinger_core_test_sources})
 target_link_libraries(stinger_core_test stinger_core gtest)
 
 #================================
 
+set(_stinger_batch_test_sources
+        stinger_batch_test/stinger_batch_test.cpp
+        stinger_batch_test/stinger_batch_test.h
+        )
+
```
Comment (ehein6) [inline src/tests/CMakeLists.txt]: Fix whitespace here too

File: src/tests/stinger_batch_test/stinger_batch_test.cpp (hunk 7 above)
Comment (ehein6) [inline src/tests/stinger_batch_test/stinger_batch_test.cpp]: Yes, added to todo list.

File: lib/stinger_core/inc/stinger_batch_insert.h (hunk 3 above)
Comment (tgoodyear) [inline lib/stinger_core/inc/stinger_batch_insert.h]: I'm ok with leaving this as is. Maybe we make a note somewhere that officially a STINGER graph is to consist entirely of either directed or undirected edges, not a combination.

**Comment by ehein6** [comment]
I did some performance testing against the old implementation to see if there was any slowdown for small batch sizes. Even as small as 5K there is no slowdown, so I'm not going to implement a fallback to the old algorithm.
image

**Comment by tgoodyear** [comment]
For the record, this branch was used to ingest over 1B edges over an 8-day period. No stability issues, processing times were consistently subsecond to a graph with |V| ~ 5M and |E| ~ 250M, batch size between 1k and 80k.

File: src/tests/stinger_batch_test/stinger_batch_test.cpp (hunk 7 above)
Comment (ehein6) [inline src/tests/stinger_batch_test/stinger_batch_test.cpp]: I've decided to omit concurrent deletion tests for now. OpenMP won't let me call `stinger_batch_incr_edges` in parallel with `stinger_remove_edge`, because this would result in a `omp parallel for` nested inside an `omp section`. To fix this, `stinger_batch` functions should handle insertions and deletions together. While this is conceptually possible, it would be quite a bit more work, which doesn't seem worthwhile at the moment because concurrent insertion/deletion isn't actually being used anywhere at the moment. Let's just get this merged in, and we can create an issue to integrate optimized batch deletion.

**Comment by ehein6** [comment]
@davidediger Pending any further comments, I believe this is ready to merge. Waiting on your approval to squash and push one more time.

**Comment by davidediger** [comment]
@ehein6 Let's go for it.

**Comment by ehein6** [comment]
@davidediger Ready to merge
